### DISCUSSION TITLE
How to configure properly TLS on Keycloak 20.0.3?

### DISCUSSION THREAD

**Discussion by gitblanc** [keycloak/keycloak#22392] [Q&A]
I'm trying to configure TLS correctly following the [official doc](https://www.keycloak.org/server/enabletls), but i always get this error:
````
Aug 11 13:35:46 localhost keycloak: 2023-08-11 13:35:46,176 INFO  [org.infinispan.CLUSTER] (main) ISPN000080: Disconnecting JGroups channel `ISPN`
Aug 11 13:35:46 localhost keycloak: 2023-08-11 13:35:46,262 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: Failed to start server in (production) mode
Aug 11 13:35:46 localhost keycloak: 2023-08-11 13:35:46,263 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) ERROR: /path_to_my_pem/file.pem
Aug 11 13:35:46 localhost keycloak: 2023-08-11 13:35:46,264 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) Key material not provided to setup HTTPS. Please configure your keys/certificates or start the server in development mode.
Aug 11 13:35:46 localhost keycloak: 2023-08-11 13:35:46,264 ERROR [org.keycloak.quarkus.runtime.cli.ExecutionExceptionHandler] (main) For more details run the same command passing the '--verbose' option. Also you can use '--help' to see the details about the usage of the particular command.
````
Any idea of what can happen?

My current keycloak configuration is this:
````
#HTTP SSL Config
https-certificate-file=/path_to_my_crt/file.crt
https-certificate-key-file=/path_to_my_key/file.key
spi-sticky-session-encoder-infinispan-should-attach-route=false
https-protocols=TLSv1.3,TLSv1.2
````

**Comment by darius-m** [comment]
Make sure that the key file is there, can be accessed by the user running the Keycloak server and that it is indeed in PEM format (a base64-encoded text file that has on the first and last lines a text starting with `------`). Also, make sure that the key can be accessed by Keycloak (it should not be encrypted, as you have not provided a password). Alternately, you could (and it is recommended) use a PKCS12 keystore. You can create a keystore using `keytool`, similar to how they use it in the [containers](https://www.keycloak.org/server/containers) guide.

**Comment by gitblanc** [comment]
So you are telling me that Keycloak runs by default without security or with a bad security? Keep in mind that I can decode a base-64 file in less than 5 seconds , so having root passwd on that file would not be "very secure"

**Reply by darius-m** [reply]
The base64 encoding is just part of how PEM files are formatted. They have the same contents as DER, but are easier to copy and paste (and through the use of `------ BEGIN/END FILE TYPE ------` delimiters, also allow you to have more than one element in the same file - e.g. a certificate + its private key); it is not meant as security mechanism.

I'm not sure what you mean by a root password on the file - private keys should be encrypted by default, be it in PEM, PKCS12, or other formats. Since symmetric encryption relies on a password to encrypt and decrypt the file, you need to tell Keycloak what that password is, otherwise it would not be able to decrypt the private key, and thus render it useless.

**Comment by Alan-lee-tw** [comment]
also want the function to configure the password to decrypt the key.
